be/src/storage/lake/tablet_manager.cpp

```diff
@@ -396,19 +399,39 @@ Status TabletManager::publish_version(int64_t tablet_id, int64_t base_version, i
 }
 
 Status apply_txn_log(const TxnLog& log, TabletMetadata* metadata) {
+    // Apply write log
     if (log.has_op_write()) {
-        if (log.op_write().has_rowset() && log.op_write().rowset().segments_size() > 0) {
+        if (log.op_write().has_rowset() && log.op_write().rowset().num_rows() > 0) {
             auto rowset = metadata->add_rowsets();
             rowset->CopyFrom(log.op_write().rowset());
             rowset->set_id(metadata->next_rowset_id());
             metadata->set_next_rowset_id(metadata->next_rowset_id() + rowset->segments_size());
         }
     }
 
+    // Apply compaction log
     if (log.has_op_compaction()) {
-        return Status::NotSupported("does not support apply compaction log yet");
+        const auto& op_compaction = log.op_compaction();
+        // Remove input rowset(s)
+        auto find_pos = metadata->rowsets().begin();
+        for (auto input_id : op_compaction.input_rowsets()) {
+            auto f = [&](const RowsetMetadataPB& r) { return r.id() == input_id; };
+            auto it = std::find_if(find_pos, metadata->rowsets().end(), f);
+            if (it == metadata->rowsets().end()) {
+                return Status::InternalError(fmt::format("input rowset {} not found", input_id));
+            }
+            find_pos = metadata->mutable_rowsets()->erase(it);
+        }
+        // Add output rowset
+        if (op_compaction.has_output_rowset() && op_compaction.output_rowset().num_rows() > 0) {
+            auto rowset = metadata->add_rowsets();
```

Comment:
If it is a unique key table, can not put the new rowset directly at the end.

Reply:
fixed